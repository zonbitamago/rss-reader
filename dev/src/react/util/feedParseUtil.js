import axios from "axios";
import Feedparser from "feedparser";
import stringToStream from "string-to-stream";

export default class feedParseUtil {
  feedparser = new Feedparser();

  feedParse(url) {
    return axios({ method: "get", url: url, timeout: 3000 })
      .then(res => {
        stringToStream(res.data).pipe(this.feedparser);
      })
      .then(() => {
        return this.parse();
      })
      .catch(e => {
        throw new Error();
      });
  }

  parse() {
    var promise = new Promise((resolve, reject) => {
      let items = [];
      this.feedparser.on("readable", function() {
        var stream = this;
        var item;
        while ((item = stream.read())) {
          items.push(item);
        }
      });

      this.feedparser.on("end", () => {
        resolve(items);
      });

      this.feedparser.on("error", err => {
        reject(err);
      });
    });

    return Promise.all([promise])
      .then(feed => {
        return feed[0];
      })
      .catch(err => {
        throw err;
      });
  }
}
