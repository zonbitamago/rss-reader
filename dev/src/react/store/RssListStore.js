import { observable, action } from "mobx";
import feedParseUtil from "../util/feedParseUtil";

class RssListStore {
  @observable rssList = [];

  @action.bound
  getRssList() {
    if (
      localStorage.getItem("rssList") == undefined ||
      localStorage.getItem("rssList").length == 0
    ) {
      this.rssList = [];
    } else {
      this.rssList = JSON.parse(localStorage.getItem("rssList"));
    }
  }

  @action.bound
  setRssList(name, url) {
    var util = new feedParseUtil();
    var ret = util.feedParse(url);
    return ret
      .then(() => {
        var duplicateList = this.rssList.filter(item => {
          if (item.name == name) {
            return item;
          }
        });

        // 重複しているものは登録しない
        if (duplicateList.length > 0) {
          return false;
        }

        this.rssList.push({ name: name, url: url });
        localStorage.setItem("rssList", JSON.stringify(this.rssList));

        return true;
      })
      .catch(() => {
        // 正しいURLではない場合、return

        return false;
      });
  }
}

export default RssListStore;
