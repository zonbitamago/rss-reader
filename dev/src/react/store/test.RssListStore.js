import RssListStore from "./RssListStore";
import feedParseUtil from "../util/feedParseUtil";
import { Promise, reject } from "bluebird-lst";

jest.mock("../util/feedParseUtil");

describe("RssListStore", () => {
  let store;

  beforeEach(() => {
    localStorage.clear();
  });

  it("init", () => {
    store = new RssListStore();
    expect(store.rssList.length).toBe(0);
  });

  describe("getRssList", () => {
    it("Not Registed", () => {
      store = new RssListStore();
      store.getRssList();
      expect(store.rssList.length).toBe(0);
    });

    it("Registed 1items", () => {
      localStorage.setItem(
        "rssList",
        JSON.stringify([
          {
            name: "google.com",
            url: "https://google.com"
          }
        ])
      );
      store = new RssListStore();
      store.getRssList();
      expect(store.rssList.length).toBe(1);
    });

    it("Registed 2items", () => {
      localStorage.setItem(
        "rssList",
        JSON.stringify([
          {
            name: "google.com",
            url: "https://google.com"
          },
          {
            name: "yahoo.com",
            url: "https://yahoo.com"
          }
        ])
      );
      store = new RssListStore();
      store.getRssList();
      expect(store.rssList.length).toBe(2);
    });
  });

  describe("setRssList", () => {
    it("Not Regsited", () => {
      feedParseUtil.mockImplementation(() => {
        return {
          feedParse: url => {
            return new Promise(resolve => {
              resolve("test");
            });
          }
        };
      });

      var name = "google.com";
      var url = "https://google.com";
      store = new RssListStore();

      var promise = store.setRssList(name, url);

      return promise.then(ret => {
        expect(ret).toBe(true);
        expect(JSON.parse(localStorage.getItem("rssList")).length).toBe(1);
        expect(JSON.parse(localStorage.getItem("rssList"))).toMatchSnapshot();
      });
    });

    it("Registed", () => {
      feedParseUtil.mockImplementation(() => {
        return {
          feedParse: url => {
            return new Promise(resolve => {
              resolve("test");
            });
          }
        };
      });

      localStorage.setItem(
        "rssList",
        JSON.stringify([
          {
            name: "google.com",
            url: "https://google.com"
          }
        ])
      );

      var name = "yahoo.com";
      var url = "https://yahoo.com";
      store = new RssListStore();
      store.getRssList();

      var promise = store.setRssList(name, url);

      return promise.then(ret => {
        expect(ret).toBe(true);
        expect(JSON.parse(localStorage.getItem("rssList")).length).toBe(2);
        expect(JSON.parse(localStorage.getItem("rssList"))).toMatchSnapshot();
      });
    });

    it("Duplicate", () => {
      feedParseUtil.mockImplementation(() => {
        return {
          feedParse: url => {
            return new Promise(resolve => {
              resolve("test");
            });
          }
        };
      });

      localStorage.setItem(
        "rssList",
        JSON.stringify([
          {
            name: "google.com",
            url: "https://google.com"
          }
        ])
      );

      var name = "google.com";
      var url = "https://google.com";
      store = new RssListStore();
      store.getRssList();

      var promise = store.setRssList(name, url);

      return promise.then(ret => {
        expect(ret).toBe(false);
        expect(JSON.parse(localStorage.getItem("rssList")).length).toBe(1);
        expect(JSON.parse(localStorage.getItem("rssList"))).toMatchSnapshot();
      });
    });

    it("Not URL", async () => {
      feedParseUtil.mockImplementation(() => {
        return {
          feedParse: url => {
            return new Promise((resolve, reject) => {
              reject(new Error());
            });
          }
        };
      });

      var name = "google.com";
      var url = "https://google.com";
      store = new RssListStore();

      var promise = store.setRssList(name, url);

      return promise.then(ret => {
        expect(ret).toBe(false);
        expect(JSON.parse(localStorage.getItem("rssList"))).toMatchSnapshot();
      });
    });
  });
});
