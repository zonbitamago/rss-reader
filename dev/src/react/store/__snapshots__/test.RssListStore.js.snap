// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RssListStore setRssList Duplicate 1`] = `
Array [
  Object {
    "name": "google.com",
    "url": "https://google.com",
  },
]
`;

exports[`RssListStore setRssList Not Regsited 1`] = `
Array [
  Object {
    "name": "google.com",
    "url": "https://google.com",
  },
]
`;

exports[`RssListStore setRssList Not URL 1`] = `null`;

exports[`RssListStore setRssList Registed 1`] = `
Array [
  Object {
    "name": "google.com",
    "url": "https://google.com",
  },
  Object {
    "name": "yahoo.com",
    "url": "https://yahoo.com",
  },
]
`;
